### PR TITLE
[CALCITE-3925] Support materialized view matching for LogicalCorrelate in Subsitutio…

### PR DESCRIPTION
**Pull request by xy2953396112**
…nVisitor (xzh)
JIRA: https://issues.apache.org/jira/browse/CALCITE-3925

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 1]
```diff
@@ -1717,6 +1907,31 @@ private static UnifyResult setOpApply(UnifyRuleCall call) {
     return null;
   }
 
+  /**
+   * Check if query equivalent to target required columns, and also check correlationId
+   */
+  private static boolean checkRequiredColumnsAndCorrelationId(MutableCorrelate query,
+      MutableCorrelate target) {
+    if (!(query.correlationId.getId() == target.correlationId.getId())) {
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:1915]: What if query and materialized view are equivalent with each other by semantics but only different by the correlationId. Should we fail the match ?

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 2]
```diff
@@ -1164,6 +1168,192 @@ public UnifyResult apply(UnifyRuleCall call) {
...
+      if (joinRelType != JoinRelType.INNER && !(joinRelType.isOuterJoin()
+          && qInput0Cond.isAlwaysTrue() && qInput1Cond.isAlwaysTrue())) {
+        return null;
+      }
+      if (!checkRequiredColumnsAndCorrelationId(query, target)) {
+        return null;
+      }
+
+      final RexNode qInput1CondShifted =
+          RexUtil.shift(qInput1Cond, fieldCnt(qInput0.getInput()));
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:1335]: What if correlate variable is referenced in condition, should we pull it up ?

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java (hunk 1 above)
Comment (xy2953396112) [reply core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:1915]: I think we should check that it's exactly the same，Maybe you can give an example. Let's discuss it together.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java (hunk 2 above)
Comment (xy2953396112) [reply core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:1335]: i think we should pull it up.

**Pull request closed without merging**